versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java

```diff
@@ -960,15 +1115,13 @@ protected Hash findCommonAncestor(OP_CONTEXT ctx, Hash from, NamedRef toBranch,
         }
       }
       if (!anyFetched) {
-        throw new ReferenceConflictException(
-            String.format(
-                "No common ancestor found for merge of '%s' into branch '%s'",
-                from, toBranch.getName()));
+        return null;
       }
 
-      for (Hash f : fromCommitHashes) {
-        if (toCommitHashes.contains(f)) {
-          return f;
+      for (int diffOnFrom = 0; diffOnFrom < fromCommitHashes.size(); diffOnFrom++) {
+        Hash f = fromCommitHashes.get(diffOnFrom);
+        if (state.contains(f)) {
+          return result.apply(diffOnFrom, f);
         }
```

Comment:
Could it be more efficient if we kept N + 1 hash sets (one for each ref, plus one for the "to" branch), fetches commits breadth-first and cross-check each commit against the other maps? I believe list searching would not be required and hopefully less transient data would be used. WDYT?

Reply:
I think, it depends on the actual branch/tag model - whether all/most branches/tags are created off of the main branch or whether those are a rather "complex graph".
My assumption here were that:
* most branches are rather new(-ish) - aka: not that far away from the default branch's HEAD
* the UI does not request the extended information for tags

Reply:
But yes, I also see that there's probably room for improvement here.

For the use case of showing the "additional information" for all branches: Git(Hub) seems to do it on the fly. At least the corresponding `git` command is pretty quick. But unlike Git we (Nessie) cannot access "everything quickly" (like just read the missing pieces off the disk).

For the use case of a merge/transplant: It's only one branch vs another branch - so probably not that much of an issue.